File: mediapipe/tasks/web/vision/hand_landmarker/hand_landmarks_connections.ts
/**
 * Copyright 2023 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Connection} from '../../../../tasks/web/vision/core/types';

/**
 * An array containing the pairs of hand landmark indices to be rendered with
 * connections.
 */
export const HAND_CONNECTIONS: Connection[] = [
  {start: 0, end: 1},   {start: 1, end: 2},   {start: 2, end: 3},
  {start: 3, end: 4},   {start: 0, end: 5},   {start: 5, end: 6},
  {start: 6, end: 7},   {start: 7, end: 8},   {start: 5, end: 9},
  {start: 9, end: 10},  {start: 10, end: 11}, {start: 11, end: 12},
  {start: 9, end: 13},  {start: 13, end: 14}, {start: 14, end: 15},
  {start: 15, end: 16}, {start: 13, end: 17}, {start: 0, end: 17},
  {start: 17, end: 18}, {start: 18, end: 19}, {start: 19, end: 20}
];

File: mediapipe/tasks/java/com/google/mediapipe/tasks/vision/handlandmarker/HandLandmarksConnections.java
// Copyright 2023 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.mediapipe.tasks.vision.handlandmarker;

import com.google.mediapipe.tasks.components.containers.Connection;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/** Hand landmarks connection constants. */
final class HandLandmarksConnections {

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_PALM_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(0, 1),
                  Connection.create(0, 5),
                  Connection.create(9, 13),
                  Connection.create(13, 17),
                  Connection.create(5, 9),
                  Connection.create(0, 17))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_THUMB_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(1, 2), Connection.create(2, 3), Connection.create(3, 4))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_INDEX_FINGER_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(5, 6), Connection.create(6, 7), Connection.create(7, 8))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_MIDDLE_FINGER_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(9, 10), Connection.create(10, 11), Connection.create(11, 12))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_RING_FINGER_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(13, 14),
                  Connection.create(14, 15),
                  Connection.create(15, 16))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_PINKY_FINGER_CONNECTIONS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  Connection.create(17, 18),
                  Connection.create(18, 19),
                  Connection.create(19, 20))));

  @SuppressWarnings("ConstantCaseForConstants")
  static final Set<Connection> HAND_CONNECTIONS =
      Collections.unmodifiableSet(
          Stream.of(
                  HAND_PALM_CONNECTIONS.stream(),
                  HAND_THUMB_CONNECTIONS.stream(),
                  HAND_INDEX_FINGER_CONNECTIONS.stream(),
                  HAND_MIDDLE_FINGER_CONNECTIONS.stream(),
                  HAND_RING_FINGER_CONNECTIONS.stream(),
                  HAND_PINKY_FINGER_CONNECTIONS.stream())
              .flatMap(i -> i)
              .collect(Collectors.toSet()));

  private HandLandmarksConnections() {}
}
